File: reference/nova-companion/core/tts-router.ts
import type { AppConfig, Character, Emotion, Env, VoiceProviderId } from './config'
import { resolveHeaders, substituteEnv } from './config'
import { geminiStylePrefix, parseEmotion, stripAllTags, ttsParams } from './emotion'

export interface TtsResult {
  audio: ArrayBuffer
  mime: string
  provider: VoiceProviderId
  voice: string
  emotion: Emotion
  text: string
  ms: number
  attempts: { provider: VoiceProviderId; status: number | string; ms: number }[]
}

export interface TtsLogEntry {
  ts: string; character: string; provider: VoiceProviderId | 'none'; voice: string; emotion: Emotion; ms: number; ok: boolean; error?: string
}

const sleep = (ms: number) => new Promise(r => setTimeout(r, ms))

// ---- ElevenLabs voice-name resolution (cached per isolate, 10 min) ------------------------
let elVoiceCache: { at: number; byName: Map<string, string> } | null = null

export async function resolveElevenVoices(cfg: AppConfig, env: Env, log: Console = console): Promise<Map<string, string>> {
  if (elVoiceCache && Date.now() - elVoiceCache.at < 10 * 60_000) return elVoiceCache.byName
  const prov = cfg.voiceRoutes.providers.elevenlabs
  const headers = resolveHeaders(prov.headers, env)
  const byName = new Map<string, string>()
  if (!headers['xi-api-key']) return byName
  try {
    const r = await fetch(prov.voicesUrl ?? 'https://api.elevenlabs.io/v1/voices', { headers })
    if (r.ok) {
      const j: any = await r.json()
      for (const v of j.voices ?? []) byName.set(String(v.name).toLowerCase(), v.voice_id)
      log.info(`[tts] ElevenLabs voices loaded: ${byName.size}`)
    } else log.warn(`[tts] ElevenLabs /v1/voices -> ${r.status}`)
  } catch (e) { log.warn(`[tts] ElevenLabs /v1/voices failed: ${(e as Error).message}`) }
  elVoiceCache = { at: Date.now(), byName }
  return byName
}

export async function elevenVoiceId(cfg: AppConfig, ch: Character, env: Env, override?: string, log: Console = console): Promise<string> {
  if (override?.trim()) { log.info(`[tts] ElevenLabs voice: user-pasted id ${override}`); return override.trim() }
  const pasted = cfg.voiceRoutes.providers.elevenlabs.userPastedId
  if (pasted?.trim()) { log.info(`[tts] ElevenLabs voice: settings-pasted id ${pasted}`); return pasted.trim() }
  const byName = await resolveElevenVoices(cfg, env, log)
  const found = byName.get(ch.voice.elevenlabs.name.toLowerCase())
  if (found) { log.info(`[tts] ElevenLabs voice "${ch.voice.elevenlabs.name}" matched by name -> ${found}`); return found }
  log.info(`[tts] ElevenLabs voice "${ch.voice.elevenlabs.name}" not found by name, using hard-coded id ${ch.voice.elevenlabs.id}`)
  return ch.voice.elevenlabs.id
}

// ---- Provider adapters --------------------------------------------------------------------
type AdapterArgs = { cfg: AppConfig; ch: Character; env: Env; text: string; emotion: Emotion; voiceOverride?: string; log: Console }
type Adapter = (a: AdapterArgs) => Promise<{ res: Response; voice: string } | { skip: string }>

const adapters: Record<VoiceProviderId, Adapter> = {
  async elevenlabs({ cfg, ch, env, text, emotion, voiceOverride, log }) {
    const prov = cfg.voiceRoutes.providers.elevenlabs
    const headers = resolveHeaders(prov.headers, env)
    if (!headers['xi-api-key']) return { skip: 'no ELEVENLABS_API_KEY' }
    const voice = await elevenVoiceId(cfg, ch, env, voiceOverride, log)
    const p = ttsParams(cfg, ch, emotion)
    const url = substituteEnv(prov.baseUrl, env).replace('{voice_id}', voice) + '?output_format=mp3_44100_128'
    const res = await fetch(url, {
      method: 'POST',
      headers: { ...headers, 'Content-Type': 'application/json' },
      body: JSON.stringify({
        text, model_id: prov.model ?? 'eleven_multilingual_v2',
        voice_settings: { stability: p.stability, style: p.style, speed: p.speed, similarity_boost: 0.8, use_speaker_boost: true },
      }),
    })
    return { res, voice }
  },

  async gemini_tts({ cfg, ch, env, text, emotion, voiceOverride }) {
    const prov = cfg.voiceRoutes.providers.gemini_tts
    const headers = resolveHeaders(prov.headers, env)
    if (!headers['x-goog-api-key']) return { skip: 'no GEMINI_API_KEY' }
    const voice = voiceOverride?.trim() || ch.voice.gemini
    const prefix = geminiStylePrefix(cfg, ch, emotion)
    const res = await fetch(substituteEnv(prov.baseUrl, env), {
      method: 'POST',
      headers: { ...headers, 'Content-Type': 'application/json' },
      body: JSON.stringify({
        contents: [{ parts: [{ text: `${prefix} ${text}` }] }],
        generationConfig: {
          responseModalities: ['AUDIO'],
          speechConfig: { voiceConfig: { prebuiltVoiceConfig: { voiceName: voice } } },
        },
      }),
    })
    return { res, voice }
  },

  async edge_tts({ cfg, ch, env, text, emotion, voiceOverride }) {
    // openai-edge-tts compatible proxy (default localhost:5050)
    const prov = cfg.voiceRoutes.providers.edge_tts
    const voice = voiceOverride?.trim() || ch.voice.edge
    const p = ttsParams(cfg, ch, emotion)
    const res = await fetch(substituteEnv(prov.baseUrl, env), {
      method: 'POST',
      headers: { ...resolveHeaders(prov.headers, env), 'Content-Type': 'application/json' },
      body: JSON.stringify({ model: 'tts-1', input: text, voice, speed: p.speed, response_format: 'mp3' }),
    })
    return { res, voice }
  },

  async kokoro_local({ cfg, ch, env, text, emotion, voiceOverride }) {
    const prov = cfg.voiceRoutes.providers.kokoro_local
    const voice = voiceOverride?.trim() || ch.voice.kokoro
    const p = ttsParams(cfg, ch, emotion)
    const res = await fetch(substituteEnv(prov.baseUrl, env), {
      method: 'POST',
      headers: { ...resolveHeaders(prov.headers, env), 'Content-Type': 'application/json' },
      body: JSON.stringify({ model: prov.model ?? 'kokoro', input: text, voice, speed: p.speed, response_format: 'mp3' }),
    })
    return { res, voice }
  },

  async openai_tts({ cfg, ch, env, text, emotion, voiceOverride }) {
    const prov = cfg.voiceRoutes.providers.openai_tts
    const headers = resolveHeaders(prov.headers, env)
    if (!headers['Authorization']) return { skip: 'no OPENAI_API_KEY' }
    const voice = voiceOverride?.trim() || ch.voice.openai
    const p = ttsParams(cfg, ch, emotion)
    const res = await fetch(substituteEnv(prov.baseUrl, env), {
      method: 'POST',
      headers: { ...headers, 'Content-Type': 'application/json' },
      body: JSON.stringify({
        model: prov.model ?? 'gpt-4o-mini-tts', input: text, voice, speed: p.speed, response_format: 'mp3',
        instructions: geminiStylePrefix(cfg, ch, emotion),
      }),
    })
    return { res, voice }
  },
}

// Gemini returns base64 PCM inside JSON; wrap in a WAV header so browsers can play it.
async function extractAudio(provider: VoiceProviderId, res: Response): Promise<{ audio: ArrayBuffer; mime: string } | null> {
  if (provider !== 'gemini_tts') {
    const buf = await res.arrayBuffer()
    if (buf.byteLength < 100) return null
    return { audio: buf, mime: res.headers.get('content-type') ?? 'audio/mpeg' }
  }
  const j: any = await res.json()
  const part = j.candidates?.[0]?.content?.parts?.find((p: any) => p.inlineData)
  if (!part) return null
  const pcm = Uint8Array.from(atob(part.inlineData.data), c => c.charCodeAt(0))
  const rateM = String(part.inlineData.mimeType ?? '').match(/rate=(\d+)/)
  return { audio: pcmToWav(pcm, rateM ? +rateM[1] : 24000), mime: 'audio/wav' }
}

function pcmToWav(pcm: Uint8Array, sampleRate: number, channels = 1, bits = 16): ArrayBuffer {
  const header = new ArrayBuffer(44); const v = new DataView(header)
  const w = (o: number, s: string) => { for (let i = 0; i < s.length; i++) v.setUint8(o + i, s.charCodeAt(i)) }
  w(0, 'RIFF'); v.setUint32(4, 36 + pcm.length, true); w(8, 'WAVE'); w(12, 'fmt ')
  v.setUint32(16, 16, true); v.setUint16(20, 1, true); v.setUint16(22, channels, true)
  v.setUint32(24, sampleRate, true); v.setUint32(28, sampleRate * channels * bits / 8, true)
  v.setUint16(32, channels * bits / 8, true); v.setUint16(34, bits, true); w(36, 'data'); v.setUint32(40, pcm.length, true)
  const out = new Uint8Array(44 + pcm.length); out.set(new Uint8Array(header)); out.set(pcm, 44)
  return out.buffer
}

// ---- Router ------------------------------------------------------------------------------
export class TTSRouter {
  readonly log: TtsLogEntry[] = []
  constructor(private cfg: AppConfig, private env: Env, private logger: Console = console) {}

  async speak(opts: { character: Character; rawReply: string; provider?: VoiceProviderId; voiceOverride?: string }): Promise<TtsResult> {
    const ch = opts.character
    const parsed = parseEmotion(opts.rawReply, ch.defaultEmotion, this.cfg.characterEngine.emotions)
    const emotion = parsed.emotion
    const text = stripAllTags(parsed.text)
    const order = opts.provider ? [opts.provider] : this.cfg.voiceRoutes.priority
    const attempts: TtsResult['attempts'] = []
    const started = Date.now()

    for (const provider of order) {
      const t0 = Date.now()
      try {
        const r = await adapters[provider]({ cfg: this.cfg, ch, env: this.env, text, emotion, voiceOverride: opts.voiceOverride, log: this.logger })
        if ('skip' in r) { attempts.push({ provider, status: `skip: ${r.skip}`, ms: 0 }); continue }
        const ms = Date.now() - t0
        if (!r.res.ok) {
          attempts.push({ provider, status: r.res.status, ms })
          this.logger.warn(`[tts] ${provider} -> ${r.res.status}, falling back`)
          await sleep(this.cfg.voiceRoutes.fallbackDelayMs); continue
        }
        const a = await extractAudio(provider, r.res)
        if (!a) { attempts.push({ provider, status: 'empty', ms }); await sleep(this.cfg.voiceRoutes.fallbackDelayMs); continue }
        attempts.push({ provider, status: 200, ms })
        this.logger.info(`[tts] served by ${provider} voice=${r.voice} emotion=${emotion} ${ms}ms`)
        this.log.push({ ts: new Date().toISOString(), character: ch.id, provider, voice: r.voice, emotion, ms, ok: true })
        return { ...a, provider, voice: r.voice, emotion, text, ms: Date.now() - started, attempts }
      } catch (e) {
        attempts.push({ provider, status: `error: ${(e as Error).message}`, ms: Date.now() - t0 })
        this.logger.warn(`[tts] ${provider} threw: ${(e as Error).message}`)
        await sleep(this.cfg.voiceRoutes.fallbackDelayMs)
      }
    }
    this.log.push({ ts: new Date().toISOString(), character: ch.id, provider: 'none', voice: '', emotion, ms: Date.now() - started, ok: false, error: JSON.stringify(attempts) })
    const err = new Error('All TTS providers failed') as Error & { attempts: typeof attempts; text: string; emotion: Emotion }
    err.attempts = attempts; err.text = text; err.emotion = emotion
    throw err
  }
}

File: reference/nova-companion/android/app/src/main/java/com/nova/companion/tts/TTSRouter.kt
package com.nova.companion.tts

import android.content.Context
import android.media.AudioAttributes
import android.media.AudioFormat
import android.media.AudioTrack
import android.media.MediaPlayer
import android.speech.tts.TextToSpeech
import android.util.Base64
import android.util.Log
import com.nova.companion.config.Character
import com.nova.companion.config.Config
import com.nova.companion.config.TtsParams
import com.nova.companion.config.substituteEnv
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.File
import java.util.concurrent.TimeUnit

data class Parsed(val emotion: String, val text: String)
data class TtsResult(val provider: String, val voice: String, val emotion: String, val text: String, val ms: Long, val file: File?)

object Emotion {
    private val TAG_RE = Regex("""^\s*(?:<emo=([a-z]+)>|\[(?:emotion:\s*)?([a-z]+)])\s*""", RegexOption.IGNORE_CASE)
    private val ALL_TAGS = Regex("""<emo=[a-z]+>|\[(?:emotion:\s*)?[a-z]+]""", RegexOption.IGNORE_CASE)

    fun parse(raw: String, fallback: String, known: List<String>): Parsed {
        val m = TAG_RE.find(raw) ?: return Parsed(fallback, raw.trim())
        val tag = (m.groupValues[1].ifEmpty { m.groupValues[2] }).lowercase()
        return Parsed(if (tag in known) tag else fallback, raw.substring(m.range.last + 1).trim())
    }
    fun strip(text: String) = ALL_TAGS.replace(text, "").replace(Regex("\\s{2,}"), " ").trim()

    fun params(cfg: Config, ch: Character, emotion: String): TtsParams {
        val base = cfg.characterEngine.emotionToTts[emotion] ?: cfg.characterEngine.emotionToTts[ch.defaultEmotion]!!
        return TtsParams(
            (base.stability + ch.bias.stability).coerceIn(0.0, 1.0),
            (base.style + ch.bias.style).coerceIn(0.0, 1.0),
            (base.speed + ch.bias.speed).coerceIn(0.7, 1.2),
        )
    }
    fun geminiPrefix(cfg: Config, ch: Character, emotion: String): String {
        val how = cfg.characterEngine.geminiStyleByEmotion[emotion] ?: cfg.characterEngine.geminiStyleByEmotion[ch.defaultEmotion]
        return "You are ${ch.name}: ${ch.style} Say this $how:"
    }
}

/**
 * ElevenLabs → Gemini TTS → Edge → Kokoro → OpenAI, then Android's built-in TextToSpeech as last resort.
 * Drops to the next provider after 401/429/empty with a 0.5 s delay. Logs which provider served.
 */
class TTSRouter(private val ctx: Context, private val cfg: Config, private val env: Map<String, String>) {
    private val http = OkHttpClient.Builder().readTimeout(30, TimeUnit.SECONDS).build()
    private val json = Json { ignoreUnknownKeys = true }
    private var elevenByName: Map<String, String>? = null
    private var systemTts: TextToSpeech? = null
    private var player: MediaPlayer? = null
    val log = ArrayDeque<String>()

    companion object { private const val TAG = "TTSRouter" }

    /** Call once on startup: GET /v1/voices and match by name; fall back to hard-coded id. */
    suspend fun warmUpElevenVoices() = withContext(Dispatchers.IO) {
        val prov = cfg.voiceRoutes.providers["elevenlabs"] ?: return@withContext
        val key = prov.headers["xi-api-key"]?.substituteEnv(env).orEmpty()
        if (key.isBlank()) { Log.i(TAG, "ElevenLabs: no key, skipping voice list"); return@withContext }
        runCatching {
            http.newCall(Request.Builder().url(prov.voicesUrl ?: "https://api.elevenlabs.io/v1/voices").header("xi-api-key", key).build()).execute().use { r ->
                if (!r.isSuccessful) { Log.w(TAG, "ElevenLabs /v1/voices -> ${r.code}"); return@use }
                elevenByName = json.parseToJsonElement(r.body!!.string()).jsonObject["voices"]!!.jsonArray
                    .associate { it.jsonObject["name"]!!.jsonPrimitive.content.lowercase() to it.jsonObject["voice_id"]!!.jsonPrimitive.content }
                Log.i(TAG, "ElevenLabs voices loaded: ${elevenByName!!.size}")
            }
        }.onFailure { Log.w(TAG, "ElevenLabs /v1/voices failed: ${it.message}") }
    }

    private fun elevenVoiceId(ch: Character, override: String?): String {
        override?.takeIf { it.isNotBlank() }?.let { Log.i(TAG, "ElevenLabs voice: user-pasted $it"); return it }
        cfg.voiceRoutes.providers["elevenlabs"]?.userPastedId?.takeIf { it.isNotBlank() }?.let { Log.i(TAG, "ElevenLabs voice: settings-pasted $it"); return it }
        elevenByName?.get(ch.voice.elevenlabs.name.lowercase())?.let { Log.i(TAG, "ElevenLabs voice '${ch.voice.elevenlabs.name}' matched by name -> $it"); return it }
        Log.i(TAG, "ElevenLabs voice '${ch.voice.elevenlabs.name}' not found by name, using hard-coded ${ch.voice.elevenlabs.id}")
        return ch.voice.elevenlabs.id
    }

    suspend fun speak(ch: Character, rawReply: String, forceProvider: String? = null, voiceOverride: String? = null): TtsResult = withContext(Dispatchers.IO) {
        val parsed = Emotion.parse(rawReply, ch.defaultEmotion, cfg.characterEngine.emotions)
        val text = Emotion.strip(parsed.text); val emotion = parsed.emotion
        val p = Emotion.params(cfg, ch, emotion)
        val order = forceProvider?.let { listOf(it) } ?: cfg.voiceRoutes.priority
        val started = System.currentTimeMillis()

        for (provider in order) {
            val prov = cfg.voiceRoutes.providers[provider] ?: continue
            val headers = prov.headers.mapValues { it.value.substituteEnv(env) }.filterValues { it.isNotBlank() && it != "Bearer " }
            val needsKey = provider in setOf("elevenlabs", "gemini_tts", "openai_tts")
            if (needsKey && headers.isEmpty()) { log("$provider skip: no key"); continue }

            val (url, body, voice) = when (provider) {
                "elevenlabs" -> { val v = elevenVoiceId(ch, voiceOverride); Triple(prov.baseUrl.replace("{voice_id}", v) + "?output_format=mp3_44100_128",
                    buildJsonObject { put("text", text); put("model_id", prov.model ?: "eleven_multilingual_v2")
                        putJsonObject("voice_settings") { put("stability", p.stability); put("style", p.style); put("speed", p.speed); put("similarity_boost", 0.8); put("use_speaker_boost", true) } }, v) }
                "gemini_tts" -> { val v = voiceOverride?.ifBlank { null } ?: ch.voice.gemini; Triple(prov.baseUrl,
                    buildJsonObject { putJsonArray("contents") { addJsonObject { putJsonArray("parts") { addJsonObject { put("text", "${Emotion.geminiPrefix(cfg, ch, emotion)} $text") } } } }
                        putJsonObject("generationConfig") { putJsonArray("responseModalities") { add("AUDIO") }
                            putJsonObject("speechConfig") { putJsonObject("voiceConfig") { putJsonObject("prebuiltVoiceConfig") { put("voiceName", v) } } } } }, v) }
                "openai_tts" -> { val v = voiceOverride?.ifBlank { null } ?: ch.voice.openai; Triple(prov.baseUrl,
                    buildJsonObject { put("model", prov.model ?: "gpt-4o-mini-tts"); put("input", text); put("voice", v); put("speed", p.speed); put("response_format", "mp3"); put("instructions", Emotion.geminiPrefix(cfg, ch, emotion)) }, v) }
                else -> { // edge_tts / kokoro_local: OpenAI-compatible /v1/audio/speech
                    val v = voiceOverride?.ifBlank { null } ?: if (provider == "edge_tts") ch.voice.edge else ch.voice.kokoro
                    Triple(prov.baseUrl, buildJsonObject { put("model", prov.model ?: "tts-1"); put("input", text); put("voice", v); put("speed", p.speed); put("response_format", "mp3") }, v) }
            }

            val t0 = System.currentTimeMillis()
            val resp = runCatching {
                http.newCall(Request.Builder().url(url).apply { headers.forEach { (k, v) -> header(k, v) } }.post(body.toString().toRequestBody("application/json".toMediaType())).build()).execute()
            }.getOrElse { log("$provider error: ${it.message}"); delay(cfg.voiceRoutes.fallbackDelayMs); continue }

            resp.use { r ->
                if (!r.isSuccessful) { log("$provider -> ${r.code}, falling back"); delay(cfg.voiceRoutes.fallbackDelayMs); return@use }
                val bytes = r.body!!.bytes()
                if (provider == "gemini_tts") {
                    val part = json.parseToJsonElement(String(bytes)).jsonObject["candidates"]?.jsonArray?.firstOrNull()?.jsonObject
                        ?.get("content")?.jsonObject?.get("parts")?.jsonArray?.firstOrNull { it.jsonObject.containsKey("inlineData") }?.jsonObject?.get("inlineData")?.jsonObject
                    if (part == null) { log("gemini_tts empty"); return@use }
                    val pcm = Base64.decode(part["data"]!!.jsonPrimitive.content, Base64.DEFAULT)
                    val rate = Regex("rate=(\\d+)").find(part["mimeType"]?.jsonPrimitive?.content ?: "")?.groupValues?.get(1)?.toInt() ?: 24000
                    playPcm(pcm, rate)
                } else {
                    if (bytes.size < 100) { log("$provider empty"); return@use }
                    val f = File.createTempFile("tts", ".mp3", ctx.cacheDir).apply { writeBytes(bytes) }
                    playFile(f)
                }
                val ms = System.currentTimeMillis() - t0
                log("served by $provider voice=$voice emotion=$emotion ${ms}ms")
                return@withContext TtsResult(provider, voice, emotion, text, System.currentTimeMillis() - started, null)
            }
        }
        // Zero keys, no local servers: Android's own TTS so the character still talks.
        log("all providers failed -> android system TTS")
        systemSpeak(ch, text, p)
        TtsResult("android_system", "default", emotion, text, System.currentTimeMillis() - started, null)
    }

    private fun playFile(f: File) { player?.release(); player = MediaPlayer().apply { setDataSource(f.path); prepare(); start() } }

    private fun playPcm(pcm: ByteArray, rate: Int) {
        val track = AudioTrack.Builder().setAudioAttributes(AudioAttributes.Builder().setUsage(AudioAttributes.USAGE_MEDIA).build())
            .setAudioFormat(AudioFormat.Builder().setEncoding(AudioFormat.ENCODING_PCM_16BIT).setSampleRate(rate).setChannelMask(AudioFormat.CHANNEL_OUT_MONO).build())
            .setBufferSizeInBytes(pcm.size).setTransferMode(AudioTrack.MODE_STATIC).build()
        track.write(pcm, 0, pcm.size); track.play()
    }

    private fun systemSpeak(ch: Character, text: String, p: TtsParams) {
        val tts = systemTts ?: TextToSpeech(ctx) { }.also { systemTts = it }
        tts.setSpeechRate(p.speed.toFloat()); tts.setPitch(if (ch.gender == "female") 1.15f else 0.9f)
        tts.speak(text, TextToSpeech.QUEUE_FLUSH, null, "nova")
    }

    private fun log(msg: String) { Log.i(TAG, msg); log.addLast(msg); if (log.size > 100) log.removeFirst() }
}
